src/vs/workbench/browser/parts/views/customView.ts

```diff
@@ -669,17 +732,21 @@ class TreeRenderer implements IRenderer {
 
 		if (resource || node.themeIcon) {
 			const fileDecorations = this.configurationService.getValue<{ colors: boolean, badges: boolean }>('explorer.decorations');
-			templateData.resourceLabel.setResource({ name: label, description, resource: resource ? resource : URI.parse('missing:_icon_resource') }, { fileKind: this.getFileKind(node), title, hideIcon: !!iconUrl, fileDecorations, extraClasses: ['custom-view-tree-node-item-resourceLabel'], matches });
+			templateData.resourceLabel.setResource({ name: label, description, resource: resource ? resource : URI.parse('missing:_icon_resource') }, { fileKind: this.getFileKind(node), title, hideIcon: !!iconUrl, fileDecorations, extraClasses: ['custom-view-tree-node-item-resourceLabel'], matches: createMatches(element.filterData) });
 		} else {
-			templateData.resourceLabel.setResource({ name: label, description }, { title, hideIcon: true, extraClasses: ['custom-view-tree-node-item-resourceLabel'], matches });
+			templateData.resourceLabel.setResource({ name: label, description }, { title, hideIcon: true, extraClasses: ['custom-view-tree-node-item-resourceLabel'], matches: createMatches(element.filterData) });
 		}
 
 		templateData.icon.style.backgroundImage = iconUrl ? `url('${iconUrl.toString(true)}')` : '';
 		DOM.toggleClass(templateData.icon, 'custom-view-tree-node-item-icon', !!iconUrl);
 		templateData.actionBar.context = (<TreeViewItemHandleArg>{ $treeViewId: this.treeViewId, $treeItemHandle: node.handle });
 		templateData.actionBar.push(this.menus.getResourceActions(node), { icon: true, label: false });
+		this.setAlignment(templateData.container, node);
+		this._register(this.themeService.onDidFileIconThemeChange(() => this.setAlignment(templateData.container, node)));
```

Comment:
This will dispose this listener only when the class is disposed. But we need to dispose this when the corresponding template is disposed. Hence, I would introduce a new property called `itemAligner` which is of type class `ItemAligner`. This new class is very similar to the `Aligner` but it aligns the given item. 

```
interface ITreeExplorerTemplateData {
	itemAligner: ItemAligner;
}
```

```
class ItemAligner extends Disposable {

	private _treeItem: ITreeItem;

	constructor(private readonly container: HTMLElement, private readonly aligner: Aligner) {
		this._register(this.themeService.onDidFileIconThemeChange(() => this.render()));
	}

	set treeItem(treeItem: ITreeItem) {
		this._treeItem = treeItem;
		this.render();
	}

	private render(): void {
		if (this._treeItem) {
			DOM.toggleClass(this.container, 'align-icon-with-twisty', this.aligner.alignIconWithTwisty(this._treeItem));
		}
	}	
}
```

And this `ItemAligner` can be disposed in `disposeTemplate`

Reply:
This is what the elementDisopsable is for :) I will use that.